consensus/cryptarchia-engine/src/ledger.rs

```diff
@@ -0,0 +1,502 @@
+use crate::{
+    crypto::Blake2b, Commitment, Config, Epoch, Header, HeaderId, LeaderProof, Nonce, Nullifier,
+    Slot,
+};
+use blake2::Digest;
+use rpds::HashTrieSet;
+use std::collections::HashMap;
+use thiserror::Error;
+
+#[derive(Clone, Debug, Error)]
+pub enum LedgerError {
+    #[error("Commitment not found in the ledger state")]
+    CommitmentNotFound,
+    #[error("Nullifier already exists in the ledger state")]
+    NullifierExists,
+    #[error("Commitment already exists in the ledger state")]
+    CommitmentExists,
+    #[error("Invalid block slot {block:?} for parent slot {parent:?}")]
+    InvalidSlot { parent: Slot, block: Slot },
+    #[error("Parent block not found: {0:?}")]
+    ParentNotFound(HeaderId),
+    #[error("Orphan block missing: {0:?}. Importing leader proofs requires the block to be validated first")]
+    OrphanMissing(HeaderId),
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct EpochState {
+    // The epoch this snapshot is for
+    epoch: Epoch,
+    // value of the ledger nonce after 'epoch_period_nonce_buffer' slots from the beginning of the epoch
+    nonce: Nonce,
+    // stake distribution snapshot taken at the beginning of the epoch
+    // (in practice, this is equivalent to the coins the are spendable at the beginning of the epoch)
+    commitments: HashTrieSet<Commitment>,
+}
+
+impl EpochState {
+    fn update_from_ledger(self, ledger: &LedgerState, config: &Config) -> Self {
+        let nonce_snapshot_slot = config.nonce_snapshot(self.epoch);
+        let nonce = if ledger.slot < nonce_snapshot_slot {
+            ledger.nonce
+        } else {
+            self.nonce
+        };
+
+        let stake_snapshot_slot = config.stake_distribution_snapshot(self.epoch);
+        let commitments = if ledger.slot < stake_snapshot_slot {
+            ledger.lead_commitments.clone()
```

Comment:
I believe you want spend_commitments here. Otherwise there is no way for new commitments to enter the "old enough" pool 
```suggestion
            ledger.spend_commitments.clone()
```